### PR TITLE
implementation to run kube acceptance tests on GKE

### PR DESCRIPTION
**Pull request by subodh1810**
Issue : https://github.com/airbytehq/airbyte/issues/4884
Tech spec: https://docs.google.com/document/d/170JgBmRJwkuGWNZxy41bkyd6z4hAkhuKbNhZtxFtaJo/edit#

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] `README.md`
    - [ ] `docs/SUMMARY.md` if it's a new connector
    - [ ] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [ ] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md` contains a reference to the new connector
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -487,3 +487,116 @@ jobs:
...
+          cluster_name: kube-acceptance-test-cluster
+          location: us-central1-c
+
+      - name: Install kubectl
+        run: |-
+          curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
+          sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
+
+      # Set up kustomize
+      - name: Set up Kustomize
```
Comment (davinchia) [inline .github/workflows/gradle.yml:568]: why do we need to install Kustomize separately? shouldn't this come with Kubectl?

File: airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java [hunk 2]
```diff
@@ -166,19 +174,26 @@
...
+          .withPassword(SOURCE_PASSWORD);
+      sourcePsql.start();
+    }
   }
 
   @AfterAll
   public static void end() {
-    sourcePsql.stop();
+    if (IS_GKE && !IS_KUBE) {
+      throw new RuntimeException("KUBE Flag should also be enabled if GKE flag is enabled");
```
Comment (davinchia) [inline airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java:188]: should this check be in the @BeforeAll method?

File: tools/bin/gke-kube-acceptance-test/acceptance_test_kube_gke.sh [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+trap "cleanUpImages && kubectl delete namespaces $NAMESPACE --grace-period=0 --force" EXIT
+
+kubectl port-forward svc/airbyte-server-svc 8001:8001 --namespace=$NAMESPACE &
+
+kubectl port-forward svc/postgres-source-svc 2000:5432 --namespace=$NAMESPACE &
+
+kubectl port-forward svc/postgres-destination-svc 3000:5432 --namespace=$NAMESPACE &
+
+sleep 10s
```
Comment (davinchia) [inline tools/bin/gke-kube-acceptance-test/acceptance_test_kube_gke.sh:68]: what is this sleep for?

File: airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/GKEPostgresConfig.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+package io.airbyte.test.acceptance;
+
+import io.airbyte.db.Database;
+import io.airbyte.db.Databases;
+import io.airbyte.test.acceptance.AcceptanceTests.Type;
+import java.util.HashMap;
+import java.util.Map;
+
+public class GKEPostgresConfig {
```
Comment (davinchia) [inline airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/GKEPostgresConfig.java:33]: can ya drop in a comment here describing that this is convenience wrapper around a postgres that exists in Kube and is mainly for acceptance test?

File: airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java [hunk 5]
```diff
@@ -166,19 +174,26 @@
 
   @BeforeAll
   public static void init() {
-    sourcePsql = new PostgreSQLContainer("postgres:13-alpine")
-        .withUsername(SOURCE_USERNAME)
-        .withPassword(SOURCE_PASSWORD);
-    sourcePsql.start();
+    if (!IS_GKE) {
```
Comment (davinchia) [inline airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java:177]: can you update the class doc comment to say that the source/dest db are actual pods when testing this on GKE?

File: airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java [hunk 6]
```diff
@@ -958,55 +1003,56 @@ private UUID getDestinationDefId() throws ApiException {
...
+    dbConfig.put("port", psql.getFirstMappedPort());
+    dbConfig.put("database", psql.getDatabaseName());
+    dbConfig.put("username", psql.getUsername());
 
-      if (withSchema) {
-        dbConfig.put("schema", "public");
-      }
-      return Jsons.jsonNode(dbConfig);
-    } catch (Exception e) {
-      throw new RuntimeException(e);
```
Comment (davinchia) [inline airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java:1008]: curious, why was this removed?

File: .github/workflows/gradle.yml [hunk 7]
```diff
@@ -487,3 +487,116 @@ jobs:
           github-token: ${{ secrets.SELF_RUNNER_GITHUB_ACCESS_TOKEN }}
           label: ${{ needs.start-kube-acceptance-test-runner.outputs.label }}
           ec2-instance-id: ${{ needs.start-kube-acceptance-test-runner.outputs.ec2-instance-id }}
+
+  start-gke-kube-acceptance-test-runner:
```
Comment (davinchia) [inline .github/workflows/gradle.yml:491]: something that will go away once we move to on demand testing - running this on every build might overload the tiny GKE cluster we have now.

File: tools/bin/gke-kube-acceptance-test/acceptance_test_kube_gke.sh [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    findAndDeleteTag "init" $TOKEN
+    findAndDeleteTag "db" $TOKEN
+    findAndDeleteTag "seed" $TOKEN
+    findAndDeleteTag "scheduler" $TOKEN
+    findAndDeleteTag "server" $TOKEN
+    findAndDeleteTag "webapp" $TOKEN
+    findAndDeleteTag "migration" $TOKEN
+}
+
+trap "cleanUpImages && kubectl delete namespaces $NAMESPACE --grace-period=0 --force" EXIT
```
Comment (davinchia) [inline tools/bin/gke-kube-acceptance-test/acceptance_test_kube_gke.sh:60]: double checking: does deleting the namespace delete all the Kube resources within?

**Review by davinchia — COMMENTED** [review]
Looks good! Some minor comments.

This is currently configured to run on every build now. One thing to call out is our tiny GKE cluster probably won't support more than 3 tests running in parallel. I think this is fine if we are going to merge in on demand testing in the next few days. If not we should probably either make this scheduled instead of every build or merge in the on-demand command with this PR.

File: tools/bin/gke-kube-acceptance-test/acceptance_test_kube_gke.sh (hunk 8 above)
Comment (subodh1810) [reply tools/bin/gke-kube-acceptance-test/acceptance_test_kube_gke.sh:60]: Yup

File: .github/workflows/gradle.yml (hunk 7 above)
Comment (subodh1810) [reply .github/workflows/gradle.yml:491]: Oh am going to add the master and cron condition in here. I have not added it cause I wanted to run this on the CI and see how it goes

File: airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java (hunk 2 above)
Comment (subodh1810) [reply airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java:188]: Ah yes!

**Comment by subodh1810** [comment]
@davinchia I wanted to run the test on the CI thats why I have not made this scheduled yet. I plan to make this scheduled and only run on master branch

> This is currently configured to run on every build now. One thing to call out is our tiny GKE cluster probably won't support more than 3 tests running in parallel. I think this is fine if we are going to merge in on demand testing in the next few days. If not we should probably either make this scheduled instead of every build or merge in the on-demand command with this PR.

File: airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java (hunk 6 above)
Comment (subodh1810) [reply airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AcceptanceTests.java:1008]: @davinchia Oh no this was not removed, I just refactored methods and broke this into two. The new method `localConfig` returns a map and the JSON conversion is done in `getDbConfig`

**Review by jrhizor — APPROVED** [review]

**Review by davinchia — APPROVED** [review]

File: tools/bin/gke-kube-acceptance-test/acceptance_test_kube_gke.sh (hunk 3 above)
Comment (subodh1810) [reply tools/bin/gke-kube-acceptance-test/acceptance_test_kube_gke.sh:68]: This is cause I observed that even after starting the port-forwarding, it was taking a couple of seconds to re-route the traffic so I put 10 seconds just to be safe
